packages/outputarea/src/data.ts

```diff
@@ -0,0 +1,295 @@
+    });
+  }
+
+  /**
+   * Set whether the model is trusted.
+   */
+  export function setTrusted(
+    loc: IOutputAreaData.DataLocation,
+    value: boolean
+  ) {
+    if (value === DatastoreExt.getField({ ...loc.record, field: 'trusted' })) {
+      return;
+    }
+    DatastoreExt.withTransaction(loc.record.datastore, () => {
+      const list = DatastoreExt.getField({
+        ...loc.record,
+        field: 'outputs'
+      });
+      DatastoreExt.updateField({ ...loc.record, field: 'trusted' }, true);
+      for (let i = 0; i < list.length; i++) {
+        const id = list[i];
+        const record = { ...loc.outputs, record: id };
+        DatastoreExt.updateField({ ...record, field: 'trusted' }, true);
+      }
+    });
+  }
+
+  /**
+   * Set the value at the specified index.
+   */
+  export function setItem(
+    loc: IOutputAreaData.DataLocation,
+    index: number,
+    value: nbformat.IOutput
+  ): void {
+    const list = DatastoreExt.getField({
+      ...loc.record,
+      field: 'outputs'
+    });
+    const trusted = DatastoreExt.getField({
+      ...loc.record,
+      field: 'trusted'
+    });
+    // Normalize stream data.
+    Private.normalize(value);
+    const record = { ...loc.outputs, record: list[index] };
+    OutputModel.fromJSON(record, value, trusted);
+  }
+
+  /**
+   * Clear all of the output.
+   */
+  export function clear(loc: IOutputAreaData.DataLocation): void {
+    const list = DatastoreExt.getField({
+      ...loc.record,
+      field: 'outputs'
+    });
+    DatastoreExt.withTransaction(loc.record.datastore, () => {
+      DatastoreExt.updateField(
+        { ...loc.record, field: 'outputs' },
+        { index: 0, remove: list.length, values: [] }
+      );
+      list.forEach(output => {
+        OutputModel.clear({ ...loc.outputs, record: output });
+      });
+    });
+  }
+
+  /**
+   * Serialize the model to JSON.
+   */
+  export function toJSON(
+    loc: IOutputAreaData.DataLocation
+  ): nbformat.IOutput[] {
+    const list = DatastoreExt.getField({
+      ...loc.record,
+      field: 'outputs'
+    });
+    return list.map(id => OutputModel.toJSON({ ...loc.outputs, record: id }));
+  }
+
+  /**
+   * Add an array of output items to the list.
+   *
+   * #### Notes
+   * This removes overwritten characters, and consolidates items if they should
+   * be combined with a previous entry rather than a new entry.
+   */
+  export function appendItem(
+    loc: IOutputAreaData.DataLocation,
+    value: nbformat.IOutput
+  ): void {
+    let trusted = DatastoreExt.getField({
+      ...loc.record,
+      field: 'trusted'
+    });
+
+    DatastoreExt.withTransaction(loc.record.datastore, () => {
+      // Normalize the value.
+      Private.normalize(value);
+
+      const list = DatastoreExt.getField({
+        ...loc.record,
+        field: 'outputs'
+      });
+
+      // Consolidate outputs if they are stream outputs of the same kind.
+      if (list.length) {
```

Comment:
I checked the performance of a overwriting progress bar here -- seemed to be good!